搜索组件若是以 “*”开头作为搜索内容站点就会报500错误

### 系统信息

- 外部访问地址: https://www.sw0.top
- 启动时间: 2024-07-15 19:01
- 版本: 2.17.2
- 构建时间: 2024-07-15 18:20
- Git Commit: aacd8c6
- Java: IBM Semeru Runtime Open Edition / 21.0.3+9-LTS
- 数据库: MySQL Community Server - GPL / 8.2.0
- 操作系统: Linux / 6.1.0-10-amd64
- 已激活主题: [Dream2.0 Plus 1.2.6.beta2](https://github.com/zsjy/halo-theme-dream2.0-plus.git)
- 已启动插件:
  - [页面静态缓存 1.0.0](https://github.com/halo-sigs/plugin-page-cache)
  - [编辑器超链接卡片 1.0.3](https://github.com/halo-sigs/plugin-editor-hyperlink-card)
  - [TinyTale小程序设置 1.0.5](https://github.com/jiewenhuang/halo-plugin-TinyTale)
  - [静态网页服务 1.0.0-alpha.4](https://www.halo.run/store/apps/app-gFkMn)
  - 联系表单 1.0.2
  - [SEO 工具集 1.0.6](https://www.halo.run/store/apps/app-FNGbT)
  - [增强备份（Backup IMproved） 1.1.1](https://www.halo.run/store/apps/app-dHakX)
  - [评论组件 2.4.0](https://github.com/halo-dev/plugin-comment-widget)
  - [朋友圈 1.3.3](https://github.com/chengzhongxue/plugin-friends)
  - [文本绘图 1.2.1](https://github.com/halo-sigs/plugin-text-diagram)
  - [Docsme 1.0.0-alpha.7](https://www.halo.run/store/apps/app-yffxw)
  - [AI 助手 1.0.0](https://www.halo.run/store/apps/app-riNgb)
  - [Umami 1.4.0](https://github.com/halo-sigs/plugin-umami)
  - [LDAP 认证 1.1.0](https://github.com/f2ccloud/plugin-ldap-login)
  - [Metrics Graph 1.0.0](https://github.com/guqing/plugin-metrics-graph)
  - [社交 IAM 认证 1.3.0](https://www.halo.run/store/apps/app-IXZkJ)
  - [Vditor 编辑器 1.7.0](https://github.com/justice2001/halo-plugin-vditor)
  - [文章导入导出 1.2.4](https://jhacker.cn/2023/halo-plugin-export2doc)
  - [对象存储（Amazon S3 协议） 1.10.0](https://github.com/halo-dev/plugin-s3)
  - [瞬间 1.6.2](https://github.com/halo-sigs/plugin-moments)
  - [微信公众号发布 1.3.1](https://www.halo.run/store/apps/app-Dtuoa)
  - [Markdown / HTML 内容块 1.2.0](https://github.com/halo-dev/plugin-hybrid-edit-block)
  - [文章限制阅读 Pro 1.3.0](https://www.halo.run/store/apps/app-TzRqd)
  - [Willow Markdown 编辑器 1.0.0](https://github.com/guqing/willow-mde)
  - [OAuth2 认证 1.2.0](https://github.com/halo-sigs/plugin-oauth2)
  - [KaTeX 2.1.0](https://github.com/halo-sigs/plugin-katex)
  - [链接管理 1.4.0](https://github.com/halo-sigs/plugin-links)
  - [live2d 看板娘 1.2.1](https://github.com/LIlGG/plugin-live2d)
  - [图库管理 1.3.0](https://github.com/halo-sigs/plugin-photos)
  - [ByteMD 1.4.0](https://github.com/ruibaby/plugin-bytemd)
  - [StackEdit 1.1.1](https://github.com/halo-sigs/plugin-stackedit)
  - [应用市场 1.3.0](https://www.halo.run/store/apps/app-VYJbF)
  - [RSS 1.3.0](https://github.com/halo-dev/plugin-feed)
  - [搜索组件 1.5.0](https://github.com/halo-dev/plugin-search-widget)
  - [Sitemap 1.1.2](https://github.com/halo-dev/plugin-sitemap)

### 使用的哪种方式运行？

Docker

### 发生了什么？

使用发现搜索内容如果以`*`开头就会报错


### 相关日志输出

```shell
2024-07-17T11:02:30.246+08:00 ERROR 7 --- [ndedElastic-183] a.w.r.e.AbstractErrorWebExceptionHandler : [d51142c9-8942]  500 Server Error for HTTP GET "/search?keyword=*"

java.lang.RuntimeException: LEADING_WILDCARD_NOT_ALLOWED: Leading wildcard is not allowed: title:*
	at run.halo.app.search.lucene.LuceneSearchEngine.search(LuceneSearchEngine.java:219) ~[classes/:2.17.2]
	Suppressed: reactor.core.publisher.FluxOnAssembly$OnAssemblyException: 
Error has been observed at the following site(s):
	*__checkpoint ⇢ Handler run.halo.search.widget.SearchView$$Lambda/0x00000000440f58b0@5b74af57 [DispatcherHandler]
	*__checkpoint ⇢ run.halo.app.security.InitializeRedirectionWebFilter [DefaultWebFilterChain]
	*__checkpoint ⇢ run.halo.app.security.LoginRequiredForAccessWebFilter [DefaultWebFilterChain]
	*__checkpoint ⇢ run.halo.app.security.device.DeviceSessionFilter [DefaultWebFilterChain]
	*__checkpoint ⇢ run.halo.app.sms.security.SmsLoginAuthenticator [DefaultWebFilterChain]
	*__checkpoint ⇢ AuthorizationWebFilter [DefaultWebFilterChain]
	*__checkpoint ⇢ ExceptionTranslationWebFilter [DefaultWebFilterChain]
	*__checkpoint ⇢ LogoutWebFilter [DefaultWebFilterChain]
	*__checkpoint ⇢ ServerRequestCacheWebFilter [DefaultWebFilterChain]
	*__checkpoint ⇢ SecurityContextServerWebExchangeWebFilter [DefaultWebFilterChain]
	*__checkpoint ⇢ LogoutPageGeneratingWebFilter [DefaultWebFilterChain]
	*__checkpoint ⇢ AnonymousAuthenticationWebFilter [DefaultWebFilterChain]
	*__checkpoint ⇢ SecurityWebFilterChainProxy [DefaultWebFilterChain]
	*__checkpoint ⇢ TotpAuthenticationFilter [DefaultWebFilterChain]
	*__checkpoint ⇢ AuthenticationWebFilter [DefaultWebFilterChain]
	*__checkpoint ⇢ SecurityWebFilterChainProxy [DefaultWebFilterChain]
	*__checkpoint ⇢ RememberMeAuthenticationFilter [DefaultWebFilterChain]
	*__checkpoint ⇢ SecurityWebFilterChainProxy [DefaultWebFilterChain]
	*__checkpoint ⇢ AuthenticationWebFilter [DefaultWebFilterChain]
	*__checkpoint ⇢ AuthenticationWebFilter [DefaultWebFilterChain]
	*__checkpoint ⇢ ReactorContextWebFilter [DefaultWebFilterChain]
	*__checkpoint ⇢ CsrfWebFilter [DefaultWebFilterChain]
	*__checkpoint ⇢ CorsWebFilter [DefaultWebFilterChain]
	*__checkpoint ⇢ HttpHeaderWriterWebFilter [DefaultWebFilterChain]
	*__checkpoint ⇢ ServerWebExchangeReactorContextWebFilter [DefaultWebFilterChain]
	*__checkpoint ⇢ org.springframework.security.web.server.WebFilterChainProxy [DefaultWebFilterChain]
	*__checkpoint ⇢ run.halo.cache.page.PageCacheWebFilter [DefaultWebFilterChain]
	*__checkpoint ⇢ run.halo.seo.tools.AdvancedRedirectWebFilter [DefaultWebFilterChain]
	*__checkpoint ⇢ run.halo.comment.widget.captcha.CommentCaptchaFilter [DefaultWebFilterChain]
	*__checkpoint ⇢ run.halo.app.sms.security.SmsLoginAuthenticator [DefaultWebFilterChain]
	*__checkpoint ⇢ run.halo.social.login.security.SocialAuthenticatorFilter [DefaultWebFilterChain]
	*__checkpoint ⇢ run.halo.oauth.Oauth2Authenticator [DefaultWebFilterChain]
	*__checkpoint ⇢ run.halo.seo.tools.CrawlRecordFilter [DefaultWebFilterChain]
	*__checkpoint ⇢ run.halo.social.login.security.SocialAuthorizationRequestRedirectWebFilter [DefaultWebFilterChain]
	*__checkpoint ⇢ run.halo.oauth.Oauth2AuthorizationRequestRedirectWebFilter [DefaultWebFilterChain]
	*__checkpoint ⇢ cc.ryanc.staticpages.endpoint.RewriteOnNotFoundFilter [DefaultWebFilterChain]
	*__checkpoint ⇢ run.halo.app.webfilter.AdditionalWebFilterChainProxy [DefaultWebFilterChain]
	*__checkpoint ⇢ HTTP GET "/search?keyword=*" [ExceptionHandlingWebHandler]
Original Stack Trace:
		at run.halo.app.search.lucene.LuceneSearchEngine.search(LuceneSearchEngine.java:219) ~[classes/:2.17.2]
		at run.halo.app.search.SearchServiceImpl.lambda$search$0(SearchServiceImpl.java:33) ~[classes/:2.17.2]
		at reactor.core.publisher.MonoSupplier.call(MonoSupplier.java:67) ~[reactor-core-3.6.7.jar:3.6.7]
		at reactor.core.publisher.FluxSubscribeOnCallable$CallableSubscribeOnSubscription.run(FluxSubscribeOnCallable.java:228) ~[reactor-core-3.6.7.jar:3.6.7]
		at reactor.core.scheduler.SchedulerTask.call(SchedulerTask.java:68) ~[reactor-core-3.6.7.jar:3.6.7]
		at reactor.core.scheduler.SchedulerTask.call(SchedulerTask.java:28) ~[reactor-core-3.6.7.jar:3.6.7]
		at java.base/java.util.concurrent.FutureTask.run(Unknown Source) ~[na:na]
		at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(Unknown Source) ~[na:na]
		at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source) ~[na:na]
		at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source) ~[na:na]
		at java.base/java.lang.Thread.run(Unknown Source) ~[na:na]
Caused by: org.apache.lucene.queryparser.flexible.core.QueryNodeException: Leading wildcard is not allowed: title:*
	at org.apache.lucene.queryparser.flexible.standard.processors.AllowLeadingWildcardProcessor.postProcessNode(AllowLeadingWildcardProcessor.java:77) ~[lucene-queryparser-9.11.1.jar:9.11.1 0c087dfdd10e0f6f3f6faecc6af4415e671a9e69 - 2024-06-23 12:31:02]
	at org.apache.lucene.queryparser.flexible.core.processors.QueryNodeProcessorImpl.processIteration(QueryNodeProcessorImpl.java:89) ~[lucene-queryparser-9.11.1.jar:9.11.1 0c087dfdd10e0f6f3f6faecc6af4415e671a9e69 - 2024-06-23 12:31:02]
	at org.apache.lucene.queryparser.flexible.core.processors.QueryNodeProcessorImpl.processChildren(QueryNodeProcessorImpl.java:112) ~[lucene-queryparser-9.11.1.jar:9.11.1 0c087dfdd10e0f6f3f6faecc6af4415e671a9e69 - 2024-06-23 12:31:02]
	at org.apache.lucene.queryparser.flexible.core.processors.QueryNodeProcessorImpl.processIteration(QueryNodeProcessorImpl.java:87) ~[lucene-queryparser-9.11.1.jar:9.11.1 0c087dfdd10e0f6f3f6faecc6af4415e671a9e69 - 2024-06-23 12:31:02]
	at org.apache.lucene.queryparser.flexible.core.processors.QueryNodeProcessorImpl.processChildren(QueryNodeProcessorImpl.java:112) ~[lucene-queryparser-9.11.1.jar:9.11.1 0c087dfdd10e0f6f3f6faecc6af4415e671a9e69 - 2024-06-23 12:31:02]
	at org.apache.lucene.queryparser.flexible.core.processors.QueryNodeProcessorImpl.processIteration(QueryNodeProcessorImpl.java:87) ~[lucene-queryparser-9.11.1.jar:9.11.1 0c087dfdd10e0f6f3f6faecc6af4415e671a9e69 - 2024-06-23 12:31:02]
	at org.apache.lucene.queryparser.flexible.core.processors.QueryNodeProcessorImpl.process(QueryNodeProcessorImpl.java:81) ~[lucene-queryparser-9.11.1.jar:9.11.1 0c087dfdd10e0f6f3f6faecc6af4415e671a9e69 - 2024-06-23 12:31:02]
	at org.apache.lucene.queryparser.flexible.standard.processors.AllowLeadingWildcardProcessor.process(AllowLeadingWildcardProcessor.java:53) ~[lucene-queryparser-9.11.1.jar:9.11.1 0c087dfdd10e0f6f3f6faecc6af4415e671a9e69 - 2024-06-23 12:31:02]
	at org.apache.lucene.queryparser.flexible.core.processors.QueryNodeProcessorPipeline.process(QueryNodeProcessorPipeline.java:78) ~[lucene-queryparser-9.11.1.jar:9.11.1 0c087dfdd10e0f6f3f6faecc6af4415e671a9e69 - 2024-06-23 12:31:02]
	at org.apache.lucene.queryparser.flexible.core.QueryParserHelper.parse(QueryParserHelper.java:219) ~[lucene-queryparser-9.11.1.jar:9.11.1 0c087dfdd10e0f6f3f6faecc6af4415e671a9e69 - 2024-06-23 12:31:02]
	at org.apache.lucene.queryparser.flexible.standard.StandardQueryParser.parse(StandardQueryParser.java:279) ~[lucene-queryparser-9.11.1.jar:9.11.1 0c087dfdd10e0f6f3f6faecc6af4415e671a9e69 - 2024-06-23 12:31:02]
	at run.halo.app.search.lucene.LuceneSearchEngine.search(LuceneSearchEngine.java:151) ~[classes/:2.17.2]
	at run.halo.app.search.SearchServiceImpl.lambda$search$0(SearchServiceImpl.java:33) ~[classes/:2.17.2]
	at reactor.core.publisher.MonoSupplier.call(MonoSupplier.java:67) ~[reactor-core-3.6.7.jar:3.6.7]
	at reactor.core.publisher.FluxSubscribeOnCallable$CallableSubscribeOnSubscription.run(FluxSubscribeOnCallable.java:228) ~[reactor-core-3.6.7.jar:3.6.7]
	at reactor.core.scheduler.SchedulerTask.call(SchedulerTask.java:68) ~[reactor-core-3.6.7.jar:3.6.7]
	at reactor.core.scheduler.SchedulerTask.call(SchedulerTask.java:28) ~[reactor-core-3.6.7.jar:3.6.7]
	at java.base/java.util.concurrent.FutureTask.run(Unknown Source) ~[na:na]
	at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(Unknown Source) ~[na:na]
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source) ~[na:na]
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source) ~[na:na]
	at java.base/java.lang.Thread.run(Unknown Source) ~[na:na]
```


### 附加信息

_No response_

--- a/application/src/main/java/run/halo/app/search/lucene/LuceneSearchEngine.java
+++ b/application/src/main/java/run/halo/app/search/lucene/LuceneSearchEngine.java
@@ -20,6 +20,7 @@ import org.apache.lucene.analysis.custom.CustomAnalyzer;
 import org.apache.lucene.analysis.standard.StandardTokenizerFactory;
 import org.apache.lucene.document.*;
 import org.apache.lucene.index.*;
+import org.apache.lucene.queryparser.classic.QueryParserBase;
 import org.apache.lucene.queryparser.flexible.core.QueryNodeException;
 import org.apache.lucene.queryparser.flexible.standard.StandardQueryParser;
 import org.apache.lucene.search.*;
@@ -142,7 +143,8 @@ public class LuceneSearchEngine implements SearchEngine, InitializingBean, Dispo
             queryParser.setFuzzyPrefixLength(FuzzyQuery.defaultPrefixLength);
 
             var keyword = option.getKeyword();
-            var query = queryParser.parse(keyword, null);
+            var escapedKeyword = QueryParserBase.escape(keyword);
+            var query = queryParser.parse(escapedKeyword, null);
             var queryBuilder = new BooleanQuery.Builder().add(query, MUST);
 
             var filterExposed = option.getFilterExposed();

--- a/application/src/test/java/run/halo/app/search/lucene/LuceneSearchEngineTest.java
+++ b/application/src/test/java/run/halo/app/search/lucene/LuceneSearchEngineTest.java
@@ -144,6 +144,20 @@ class LuceneSearchEngineTest {
         assertEquals("<fake-tag>fake</fake-tag>-content", gotHaloDoc.getContent());
     }
 
+    @Test
+    void shouldNotThrowExceptionWhenKeywordContainsLeadingWildcard() {
+        this.searchEngine.addOrUpdate(List.of(createFakeHaloDoc()));
+
+        var option = new SearchOption();
+        option.setLimit(10);
+
+        option.setKeyword("*");
+        assertDoesNotThrow(() -> this.searchEngine.search(option));
+
+        option.setKeyword("*fake");
+        assertDoesNotThrow(() -> this.searchEngine.search(option));
+    }
+
     void runConcurrently(Runnable runnable) throws ExecutionException, InterruptedException, TimeoutException {
         var executorService = Executors.newFixedThreadPool(10);
         var futures = IntStream.of(0, 10)

